File: projects/epc/playground/src/tools/TextSplitter.cpp
#include "TextSplitter.h"
#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/classification.hpp>
#include <proxies/hwui/Font.h>

TextSplitter::TextSplitter(std::shared_ptr<Font> font, double width, const Glib::ustring &text)
{
  std::list<Glib::ustring> newLined;
  boost::split(newLined, text, boost::is_any_of("\n"));

  for(Glib::ustring line : newLined)
  {
    boost::split(words, line, boost::is_any_of(" "));

    while(!words.empty())
      lines.push_back(eatWords(font, width, words));
  }
}
Glib::ustring TextSplitter::eatWords(std::shared_ptr<Font> font, double width, std::list<Glib::ustring> &words)
{
  Glib::ustring lineText = "";
  int spaceWidth = font->getStringWidth(" ");

  while(!words.empty())
  {
    Glib::ustring word = words.front();
    words.pop_front();

    int w = 0;

    w = font->getStringWidth(word);

    if(w < width)
    {
      if(!lineText.empty())
        lineText += " ";

      lineText += word;
      width -= w;
      width -= spaceWidth;
    }
    else
    {
      if(lineText.empty())
      {
        int half = word.length() / 2;

        if(half == 0)
        {
          if(!lineText.empty())
            lineText += " ";

          lineText += word;
          width -= w;
          width -= spaceWidth;
        }
        else
        {
          Glib::ustring first = word.substr(0, half);
          Glib::ustring second = word.substr(half);
          words.push_front(second);
          words.push_front(first);
        }
      }
      else
      {
        words.push_front(word);
        break;
      }
    }
  }

  return lineText;
}

const std::list<Glib::ustring> &TextSplitter::getLines()
{
  return lines;
}

File: projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/world/overlay/TextSplitter.java
package com.nonlinearlabs.client.world.overlay;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;

import com.google.gwt.canvas.dom.client.Context2d;

public class TextSplitter {
	private LinkedList<String> words;
	private ArrayList<String> lines = new ArrayList<String>();
	private boolean widthToSmall = false;

	public TextSplitter(Context2d ctx, double width, String text) {
		String[] newLined = text.split("\\n");

		for (String line : newLined) {
			words = new LinkedList<String>(Arrays.asList(line.split("\\s")));

			while (!words.isEmpty())
				lines.add(eatWords(ctx, width, words));
		}
	}

	public ArrayList<String> getLines() {
		return lines;
	}

	private String eatWords(Context2d ctx, double width, LinkedList<String> words) {
		String lineText = "";
		double spaceWidth = ctx.measureText(" ").getWidth();

		while (!words.isEmpty()) {
			String word = words.removeFirst();
			double w = ctx.measureText(word).getWidth();
			if (w < width) {
				if (!lineText.isEmpty())
					lineText += " ";
				lineText += word;
				width -= w;
				width -= spaceWidth;
			} else {
				if (lineText.isEmpty()) {
					int half = word.length() / 2;
					if (half == 0) {
						if (!lineText.isEmpty())
							lineText += " ";
						lineText += word;
						width -= w;
						width -= spaceWidth;
					} else {
						String first = word.substring(0, half);
						String second = word.substring(half);
						words.addFirst(second);
						words.addFirst(first);
						widthToSmall = true;
					}
				} else {
					words.addFirst(word);
					break;
				}
			}
		}

		return lineText;
	}

	public boolean widthToSmall() {
		return widthToSmall;
	}
}
